### ISSUE TITLE
Aggregate and vend comments upon request (Lexer).

### ISSUE THREAD

**Issue by jordwalke** [whitequark/ocaml-m17n#3]
I was experimenting with using this as a drop in replacement for the standard OCaml lexer for an experimental (but potentially IDE grade) source pretty printer that performs line wrapping _and_ preserves comments. I ran into a tiny issue because I `m17n`'s lexer doesn't vend comments as far as I can tell.

**Comment by whitequark** [comment]
Not only this isn't a bug, but I'm not even sure what kind of API do you want. I suggest you implement it yourself, given that the only thing you need is to take `[loc]` [here](https://github.com/whitequark/ocaml-m17n/blob/master/src/m17n_lexer.ml#L560) and perform some semantic action.

**Comment by jordwalke** [comment]
Correct, this is not a bug. Is there a better place for general discussion/features other than issues? Most projects use github issues to track features or to resolve discussions as well as bugs. If you think it's a stupid idea, please feel free to just close it as #wontfix.

My idea was to have the `m17n` lexer interface match the ocaml `Lexer` module's interface as much as possible including the `comments` structure item - which vends comments.

https://github.com/ocaml/ocaml/blob/trunk/parsing/lexer.mll#L667

I will take a look at implementing it.

**Comment by whitequark** [comment]
No, I didn't mean that it's an inappropriate place to discuss, just that I don't have motivation to implement it. Your idea works.

**Comment by whitequark** [comment]
@jordwalke do you still want/need this?

**Comment by jordwalke** [comment]
We don't need this much at the moment. We extracted out the comments using the stock OCaml lexer. I am, however, interested in eventually bringing in m17n into Reason so that enough people can have it out of the box, such that m17n can gain traction. My hunch is that you need enough people testing everything in the custom rtop, and writing libraries with m17n functionality to be able to make it actually take off. Reason could be a good place to make sure enough people get it out of the box. I'm not sure if you think that's a good idea, but since we already have our own custom rtop built, merlin parser integration, and example for all the major build systems, it's a nice place to have m17n piggy back.
We do have some fairly invasive syntax changes coming up, so maybe it would be best to wait about a month to see what that amounts to. Never hurts to investigate the opportunity though.

Feel free to close this if you don't plan on implementing it otherwise.

**Comment by whitequark** [comment]
> Reason could be a good place to make sure enough people get it out of the box. I'm not sure if you think that's a good idea, but since we already have our own custom rtop built, merlin parser integration, and example for all the major build systems, it's a nice place to have m17n piggy back.

The plan sounds good to me. I will note that I will not have time to do much in course of actually implementing that.
